File: web/studio/test/default-file.test.ts
// T138 ⟨R1⟩ — the TS lane of the shared default-file contract. Reads the SAME canonical vectors the Go
// lane runs (docs/contracts/default-file.vectors.json) directly, so the two cannot describe "default"
// differently: change either lane's predicate without changing the vectors and that lane goes red. (Both
// lanes reading the one file is stronger than a mirror+diff — there is no second copy to drift.)
import { describe, it, expect } from "vitest";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { defaultFile } from "../src/defaultFile";

type Case = {
  name: string;
  files: { filename: string; contentType: string; displayOrder: number }[];
  expected: string | null;
};

const vectors: { cases: Case[] } = JSON.parse(
  readFileSync(fileURLToPath(new URL("../../../docs/contracts/default-file.vectors.json", import.meta.url)), "utf8"),
);

describe("default-file contract (T138 ⟨R1⟩)", () => {
  it("has cases", () => expect(vectors.cases.length).toBeGreaterThan(0));
  for (const c of vectors.cases) {
    it(c.name, () => {
      expect(defaultFile(c.files)?.filename ?? null).toEqual(c.expected);
    });
  }
});

File: core/internal/app/defaultfile_test.go
package app_test

import (
	"encoding/json"
	"os"
	"testing"

	"troubastack/core/internal/app"
)

// TestDefaultFileVectors runs the shared T138 ⟨R1⟩ contract (docs/contracts/default-file.vectors.json).
// The SAME file is mirrored into web/studio and run by TS; ci.yml diffs the two copies, so a change to
// the rule in one lane that isn't matched in the other fails. Teeth: flip the PDF predicate or drop the
// filename tiebreak in DefaultFile and a case here goes red.
func TestDefaultFileVectors(t *testing.T) {
	raw, err := os.ReadFile("../../../docs/contracts/default-file.vectors.json")
	if err != nil {
		t.Fatalf("read vectors: %v", err)
	}
	var doc struct {
		Cases []struct {
			Name     string         `json:"name"`
			Files    []app.SongFile `json:"files"`
			Expected *string        `json:"expected"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatalf("parse vectors: %v", err)
	}
	if len(doc.Cases) == 0 {
		t.Fatal("no cases in vectors")
	}
	for _, c := range doc.Cases {
		got, ok := app.DefaultFile(c.Files)
		switch {
		case c.Expected == nil && ok:
			t.Errorf("%s: got default %q, want none", c.Name, got.Filename)
		case c.Expected != nil && !ok:
			t.Errorf("%s: got none, want %q", c.Name, *c.Expected)
		case c.Expected != nil && ok && got.Filename != *c.Expected:
			t.Errorf("%s: got %q, want %q", c.Name, got.Filename, *c.Expected)
		}
	}
}
